_posts/2021/2021-02-22-url-helpers-outside-views-controllers.md

```diff
@@ -0,0 +1,89 @@
+---
+title: "Use Rails URL helpers outside view and controllers"
+description: "Sometimes useful to have in a job or model"
+layout: article
+category: ruby
+image:
+  base: "2021/url-helpers-outside-views-controllers"
+  alt: "Abstract greyscale roads"
+  source: "https://unsplash.com/photos/36b7JBzhfF4"
+  credit: "Bogdan Karlenko"
+---
+
+Occasionally you'll need to generate a URL for your own application outside of the normal places; views or controllers.
+
+
+## Instead of ...
+
+...hard coding a URL:
+
+```ruby
+class RequestUserCallBackJob < ApplicationJob
+  def perform(user)
+    Net::HTTP.post(
+      "http://userinfoapi.com/",
+      body: {callback_to: "https://myapp.com/user/#{user.id}"})
+  end
+end
+```
+
+
+## Include...
+
+...the route helpers that Rails includes automatically in controllers.
+
+```ruby
+class RequestUserCallBackJob < ApplicationJob
+  include Rails.application.routes.url_helpers
+
+  def perform(user)
+    Net::HTTP.post(
+      "http://userinfoapi.com/",
+      body: {callback_to: user_url(user, host: "myapp.com")})
+  end
+end
+```
+
+You'll need to ensure you specify a `host` for any URLs you generate as the URL helpers are expecting the context of an HTTP request (which is always available in the views and controllers) which sets that option.
```

Comment:
This is confusing. There is definitely a simpler way to say this!

Reply:
```suggestion
URL helpers are normally used in the context of a web request. The view or controller action will be provided the `host` or domain by the application. Outside of this context you'll need to ensure you specify a `host` for any helpers ending in `_url`.
```